experiment/manual-trigger/manual-trigger.go

```diff
@@ -0,0 +1,190 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// manual-trigger triggers jenkins jobs based a specified github pull request
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/url"
+	"strings"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/jenkins"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type options struct {
+	githubapiurl string
+	num          int
+	org          string
+	repo         string
+	ghtoken      string
+	jbtoken      string
+	jenkinsurl   string
+	jobname      string
+}
+
+func flagOptions() options {
+	o := options{}
+	flag.StringVar(&o.githubapiurl, "githubapiurl", "https://api.github.com", "GitHub's API endpoint.")
+	flag.StringVar(&o.jenkinsurl, "jenkinsurl", "", "Jenkins url.")
+
+	flag.StringVar(&o.ghtoken, "ghtoken", "", "Path to file containing GitHub token.")
+	flag.StringVar(&o.jbtoken, "jbtoken", "", "Path to file containing Jenkins Bearer token.")
+
+	flag.StringVar(&o.jobname, "jobname", "", "Name of Jenkins job")
+
+	flag.IntVar(&o.num, "num", 0, "GitHub issue number")
+	flag.StringVar(&o.org, "org", "", "GitHub organization")
+	flag.StringVar(&o.repo, "repo", "", "GitHub repository")
+	flag.Parse()
+	return o
+}
+
+func sanityCheckFlags(o options) error {
+	if o.num <= 0 {
+		return fmt.Errorf("empty or invalid --num")
+	}
+	if o.repo == "" {
+		return fmt.Errorf("empty --org")
+	}
+	if o.repo == "" {
+		return fmt.Errorf("empty --repo")
+	}
+	if o.ghtoken == "" {
+		return fmt.Errorf("empty --ghtoken")
+	}
+	if o.jbtoken == "" {
+		return fmt.Errorf("empty --jbtoken")
+	}
+	if o.jbtoken == "" {
+		return fmt.Errorf("empty --jobname")
+	}
+
+	_, err := url.Parse(o.githubapiurl)
+	if o.githubapiurl == "" || err != nil {
+		return fmt.Errorf("Must specify a valid URL to --githubapiurl")
+	}
+
+	_, err = url.Parse(o.jenkinsurl)
+	if o.jenkinsurl == "" || err != nil {
+		return fmt.Errorf("Must specify a valid URL to --jenkinsurl")
+	}
+
+	return nil
+}
+
+func readToken(path string) (string, error) {
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(string(contents)), nil
+}
+
+func main() {
+	o := flagOptions()
+	err := sanityCheckFlags(o)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ghtoken, err := readToken(o.ghtoken)
+	if err != nil {
+		log.Fatal("cannot read file specified by --ghtoken: %v", err)
+	}
+
+	jbtoken, err := readToken(o.jbtoken)
+	if err != nil {
+		log.Fatal("cannot read file specified by --jbtoken: %v", err)
+	}
+
+	c := github.NewClient(ghtoken, o.githubapiurl)
```

Comment:
Try to hoist as much as possible before this call -- the Jenkins client set-up, etc -- so we don't use tokens that would get wasted when something else down the line fails

Reply:
The only things I see that could fail after this depend on it:

- gc.GetPullRequest can fail, but needs gc
- gc.GetRef can fail, but it needs gc
- jc.Build can fail, but it needs gc.GetRef which needs gc

I can move jenkins.NewClient() up before there, but I don't see the benefit.  Not sure if I'm missing something.